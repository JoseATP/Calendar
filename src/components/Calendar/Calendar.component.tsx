/* eslint-disable react-hooks/exhaustive-deps */
import { useState, useEffect, FC } from "react";
import {
  Frame,
  Header,
  Button,
  BulletButton,
  Day,
  Body,
  SelectedMonth,
  SelectedYear,
  Footer,
} from "./Calendar.styles";
import { TReminder } from "../../interface/reminder.interface";

type CalendarProps = {
  reminders: TReminder[];
  updateFilteredReminders?: any;
  updateSelectedDate?: any;
};

export const Calendar: FC<CalendarProps> = ({
  reminders,
  updateFilteredReminders,
  updateSelectedDate,
}) => {
  const DAYS = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  const DAYS_LEAP = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  const DAYS_OF_THE_WEEK = ["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"];
  const MONTHS = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "Octuber",
    "November",
    "December",
  ];

  const today = new Date();
  const [date, setDate] = useState(today);
  const [day, setDay] = useState(date.getDate());
  const [month, setMonth] = useState(date.getMonth());
  const [year, setYear] = useState(date.getFullYear());
  const [startDay, setStartDay] = useState(getStartDayOfMonth(date));

  useEffect(() => {
    setDay(date.getDate());
    setMonth(date.getMonth());
    setYear(date.getFullYear());
    setStartDay(getStartDayOfMonth(date));
  }, [date]);

  function getStartDayOfMonth(date: Date) {
    return new Date(date.getFullYear(), date.getMonth(), 1).getDay();
  }

  function isLeapYear(year: number) {
    return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
  }

  const days = isLeapYear(date.getFullYear()) ? DAYS_LEAP : DAYS;

  const hasReminder = ({
    year,
    month,
    d,
  }: {
    year: number;
    month: number;
    d: number;
  }) => {
    const remindersCopy = [...reminders];
    const dateToCheck = new Date(year, month, d).toISOString().split("T")[0];
    const hasReminder = remindersCopy.find(
      (reminder) => reminder.date === dateToCheck
    );

    if (hasReminder) {
      return true;
    }

    return false;
  };

  const handleNewDate = (newDate: Date) => {
    setDate(newDate);
    updateSelectedDate(newDate);
  };

  useEffect(() => {
    if (updateFilteredReminders) {
      const remindersFiltered = reminders.filter(
        (reminder) => reminder.date === date.toISOString().split("T")[0]
      );
      updateFilteredReminders(remindersFiltered);
    }
  }, [date, reminders]);

  return (
    <Frame>
      <SelectedYear>{year}</SelectedYear>
      <SelectedMonth>{MONTHS[month]}</SelectedMonth>
      <Header>
        <Button onClick={() => handleNewDate(new Date(year, month - 1, day))}>
          {"<"}
        </Button>
        <BulletButton onClick={() => handleNewDate(new Date())}>{"â€¢"}</BulletButton>
        <Button onClick={() => handleNewDate(new Date(year, month + 1, day))}>
          {">"}
        </Button>
      </Header>
      <Body>
        {DAYS_OF_THE_WEEK.map((d) => (
          <Day key={d}>
            <strong>{d}</strong>
          </Day>
        ))}
        {Array(days[month] + (startDay - 1))
          .fill(null)
          .map((_, index) => {
            const d = index - (startDay - 2);
            return (
              <Day
                key={index}
                data-testid={`${d}`}
                isToday={
                  d === today.getDate() &&
                  month === today.getMonth() &&
                  year === today.getFullYear()
                }
                isSelected={d === day}
                onClick={() => {
                  setDate(new Date(year, month, d));
                  updateSelectedDate(new Date(year, month, d));
                }}
                hasReminder={hasReminder({ year, month, d })}
                day={d}
              >
                {d > 0 ? d : ""}
              </Day>
            );
          })}
      </Body>
      <Footer>@ 2020 Codelitt Inc All rights reserved</Footer>
    </Frame>
  );
};
