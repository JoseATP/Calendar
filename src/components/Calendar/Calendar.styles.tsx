import styled, { css } from "styled-components";

export const Frame = styled.div`
  width: 100%;
  padding: 40px 0px;
`;

export const SelectedMonth = styled.div`
  text-align: center;
  font-family: "Open Sans", sans-serif;
  font-weight: bolder;
  font-size: 50px;
`;

export const SelectedYear = styled.div`
  text-align: center;
  font-family: "Open Sans", sans-serif;
  font-weight: bolder;
  font-size: 80px;
`;

export const Header = styled.div`
  font-size: 18px;
  font-weight: bold;
  padding: 10px 10px 5px 10px;
  display: flex;
  justify-content: center;
`;

export const Button = styled.div`
  cursor: pointer;
`;

export const BulletButton = styled.div`
  cursor: pointer;
  margin: 0px 5px 0px 5px
`;

export const Body = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
`;

type DayProps = {
  isToday?: boolean;
  isSelected?: boolean;
  hasReminder?: boolean;
  day?: number | string;
};

export const Day = styled.div<DayProps>`
  width: 14.2%;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  position: relative;
  z-index: 1;

  ${(props) =>
    props.isToday &&
    css`
      ::after {
        content: "";
        display: flex;
        align-items: center;
        justify-content: center;
        position: absolute;
        top: 5px;
        left: 18px;
        width: 30px;
        height: 30px;
        border-radius: 100%;
        background: linear-gradient(90deg, #ec9f05 0%, #ff4e00 100%);
        z-index: -1;
      }
    `}

  ${(props) =>
    props.isSelected &&
    css`
      ::after {
        content: "";
        display: flex;
        align-items: center;
        justify-content: center;
        position: absolute;
        top: 5px;
        left: 18px;
        width: 30px;
        height: 30px;
        border-radius: 100%;
        background: linear-gradient(90deg, #ff465d 0%, #bc46ba 100%);
        z-index: -1;
      }
    `}

  ${(props) =>
    props.hasReminder &&
    css`
      ::before {
        content: "";
        position: absolute;
        top: 10px;
        right: 20px;
        display: inline-block;
        width: 5px;
        height: 5px;
        border-radius: 100%;
        background: #00ffcc;
        z-index: 1;
      }
    `}
`;

export const Footer = styled.footer`
  font-weight: 300;
  margin: 70px 0px;
  text-align: center;
`;
