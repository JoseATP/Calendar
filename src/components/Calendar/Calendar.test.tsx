import { render, fireEvent, screen, cleanup } from "@testing-library/react";
import { remindersData } from "../../mocks/reminders.mock";
import { Calendar } from "./Calendar.component";

afterEach(cleanup);

describe("Calendar", () => {
  const updateFilteredReminders = jest.fn();
  const updateSelectedDate = jest.fn();

  it("Should render Calendar component correctly", () => {
    const { asFragment } = render(
      <Calendar
        reminders={remindersData}
        updateFilteredReminders={updateFilteredReminders}
        updateSelectedDate={updateSelectedDate}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Check onSelect date function call", () => {
    render(
      <Calendar
        reminders={remindersData}
        updateFilteredReminders={updateFilteredReminders}
        updateSelectedDate={updateSelectedDate}
      />
    );

    const Day = screen.getByTestId("1");

    fireEvent.click(Day);

    expect(updateFilteredReminders).toHaveBeenCalledTimes(2);
    expect(updateSelectedDate).toHaveBeenCalledTimes(1);
  });
});
