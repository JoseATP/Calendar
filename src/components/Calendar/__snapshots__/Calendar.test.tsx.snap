// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Should render Calendar component correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beySbM boVawX"
  >
    <div
      class="sc-dmyDGi iCdgqr"
    >
      2023
    </div>
    <div
      class="sc-guDLRT cjNhkc"
    >
      March
    </div>
    <div
      class="sc-hLQRIN jEgeFZ"
    >
      <div
        class="sc-eDLJxc hueVUY"
      >
        &lt;
      </div>
      <div
        class="sc-jTQDnj cdamwV"
      >
        â€¢
      </div>
      <div
        class="sc-eDLJxc hueVUY"
      >
        &gt;
      </div>
    </div>
    <div
      class="sc-gLLvby fEeyzK"
    >
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          MON
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          TUE
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          WED
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          THU
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          FRI
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          SAT
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
      >
        <strong>
          SUN
        </strong>
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="-1"
      />
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="0"
      />
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="1"
      >
        1
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="2"
      >
        2
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="3"
      >
        3
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="4"
      >
        4
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="5"
      >
        5
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="6"
      >
        6
      </div>
      <div
        class="sc-iBdmCd kYlqSU"
        data-testid="7"
      >
        7
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="8"
      >
        8
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="9"
      >
        9
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="10"
      >
        10
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="11"
      >
        11
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="12"
      >
        12
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="13"
      >
        13
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="14"
      >
        14
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="15"
      >
        15
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="16"
      >
        16
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="17"
      >
        17
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="18"
      >
        18
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="19"
      >
        19
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="20"
      >
        20
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="21"
      >
        21
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="22"
      >
        22
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="23"
      >
        23
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="24"
      >
        24
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="25"
      >
        25
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="26"
      >
        26
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="27"
      >
        27
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="28"
      >
        28
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="29"
      >
        29
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="30"
      >
        30
      </div>
      <div
        class="sc-iBdmCd iNnggi"
        data-testid="31"
      >
        31
      </div>
    </div>
    <footer
      class="sc-fsYeqs kLtofk"
    >
      @ 2020 Codelitt Inc All rights reserved
    </footer>
  </div>
</DocumentFragment>
`;
